### ISSUE TITLE
IPv6 Firewall: no incoming connection to mailcow container

### ISSUE THREAD

**Issue by niklasdoerfler** [mailcow/mailcow-dockerized#345]
Hi,
currently I'm trying to setup IPv6 support on my mailcow docker container. On my host system I'm using iptables to secure the server by blocking all incoming connections and only open neccessary ports. With IPv4 everything is working fine, but I can't get a working incoming IPv6 connection. 
I'm using the latest mailcow release. When I remove all my own rules and start the docker container afterwards, it works. Below my ip6tables rules.

Does anyone know, which rule is missing?
Thanks in advance!
Niklas

```
# Generated by ip6tables-save v1.6.1 on Fri Jun  9 20:19:26 2017
*nat
:PREROUTING ACCEPT [49005:4435767]
:INPUT ACCEPT [290:21839]
:OUTPUT ACCEPT [737:59503]
:POSTROUTING ACCEPT [28714:2299903]
:DOCKER - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT ! -d ::1/128 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s fd4d:6169:6c63:6f77::/64 ! -o br-0980a49f53f2 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 993 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 995 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 110 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 143 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 4190 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::b/128 -d fd4d:6169:6c63:6f77::b/128 -p tcp -m tcp --dport 25 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::b/128 -d fd4d:6169:6c63:6f77::b/128 -p tcp -m tcp --dport 465 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::b/128 -d fd4d:6169:6c63:6f77::b/128 -p tcp -m tcp --dport 587 -j MASQUERADE
-A DOCKER -i br-0980a49f53f2 -j RETURN
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 993 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:993
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 995 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:995
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 110 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:110
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 143 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:143
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 4190 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:4190
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 25 -j DNAT --to-destination [fd4d:6169:6c63:6f77::b]:25
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 465 -j DNAT --to-destination [fd4d:6169:6c63:6f77::b]:465
-A DOCKER ! -i br-0980a49f53f2 -p tcp -m tcp --dport 587 -j DNAT --to-destination [fd4d:6169:6c63:6f77::b]:587
COMMIT
# Completed on Fri Jun  9 20:19:26 2017
# Generated by ip6tables-save v1.6.1 on Fri Jun  9 20:19:26 2017
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [299332:200935435]
:DOCKER - [0:0]
:DOCKER-ISOLATION - [0:0]
:TCP - [0:0]
:UDP - [0:0]
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -i tun0 -j ACCEPT
-A INPUT -m conntrack --ctstate INVALID -j DROP
-A INPUT -s fe80::/10 -p ipv6-icmp -j ACCEPT
-A INPUT -p ipv6-icmp -m icmp6 --icmpv6-type 128 -m conntrack --ctstate NEW -j ACCEPT
-A INPUT -p udp -m conntrack --ctstate NEW -j UDP
-A INPUT -p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m conntrack --ctstate NEW -j TCP
-A INPUT -p udp -j REJECT --reject-with icmp6-port-unreachable
-A INPUT -p tcp -j REJECT --reject-with tcp-reset
-A INPUT -j REJECT --reject-with icmp6-port-unreachable
-A FORWARD -j DOCKER-ISOLATION
-A FORWARD -j ACCEPT
-A FORWARD -i tun0 -j ACCEPT
-A FORWARD -o br-0980a49f53f2 -j DOCKER
-A FORWARD -o br-0980a49f53f2 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i br-0980a49f53f2 ! -o br-0980a49f53f2 -j ACCEPT
-A FORWARD -i br-0980a49f53f2 -o br-0980a49f53f2 -j ACCEPT
-A FORWARD -i ens3 -o docker0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 993 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 995 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 110 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 143 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 4190 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::b/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 25 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::b/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 465 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::b/128 ! -i br-0980a49f53f2 -o br-0980a49f53f2 -p tcp -m tcp --dport 587 -j ACCEPT
-A DOCKER-ISOLATION -j RETURN
-A TCP -p tcp -m tcp --dport 22 -j ACCEPT
-A TCP -p tcp -m tcp --dport 80 -j ACCEPT
-A TCP -p tcp -m tcp --dport 443 -j ACCEPT
-A UDP -p udp -m udp --dport 1194 -j ACCEPT
COMMIT
```

**Comment by andristeiner** [comment]
I have a working setup in a similar setting, maybe you can compore our rules to each other to find the corresponding one.

```
# Generated by ip6tables-save v1.4.21 on Fri Jun  9 20:53:04 2017
*nat
:PREROUTING ACCEPT [220518:17809528]
:INPUT ACCEPT [43512:3649048]
:OUTPUT ACCEPT [7514:601520]
:POSTROUTING ACCEPT [184521:14762080]
:DOCKER - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT ! -d ::1/128 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s fd4d:6169:6c63:6f77::/64 ! -o br-938897f8afb4 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::7/128 -d fd4d:6169:6c63:6f77::7/128 -p tcp -m tcp --dport 993 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::7/128 -d fd4d:6169:6c63:6f77::7/128 -p tcp -m tcp --dport 995 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::7/128 -d fd4d:6169:6c63:6f77::7/128 -p tcp -m tcp --dport 110 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::7/128 -d fd4d:6169:6c63:6f77::7/128 -p tcp -m tcp --dport 143 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::7/128 -d fd4d:6169:6c63:6f77::7/128 -p tcp -m tcp --dport 4190 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::8/128 -d fd4d:6169:6c63:6f77::8/128 -p tcp -m tcp --dport 25 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::8/128 -d fd4d:6169:6c63:6f77::8/128 -p tcp -m tcp --dport 465 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::8/128 -d fd4d:6169:6c63:6f77::8/128 -p tcp -m tcp --dport 587 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 587 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 25 -j MASQUERADE
-A POSTROUTING -s fd4d:6169:6c63:6f77::9/128 -d fd4d:6169:6c63:6f77::9/128 -p tcp -m tcp --dport 465 -j MASQUERADE
-A DOCKER -i br-938897f8afb4 -j RETURN
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 995 -j DNAT --to-destination [fd4d:6169:6c63:6f77::7]:995
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 110 -j DNAT --to-destination [fd4d:6169:6c63:6f77::7]:110
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 143 -j DNAT --to-destination [fd4d:6169:6c63:6f77::7]:143
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 4190 -j DNAT --to-destination [fd4d:6169:6c63:6f77::7]:4190
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 993 -j DNAT --to-destination [fd4d:6169:6c63:6f77::7]:993
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 587 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:587
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 25 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:25
-A DOCKER ! -i br-938897f8afb4 -p tcp -m tcp --dport 465 -j DNAT --to-destination [fd4d:6169:6c63:6f77::9]:465
COMMIT
# Completed on Fri Jun  9 20:53:04 2017
# Generated by ip6tables-save v1.4.21 on Fri Jun  9 20:53:04 2017
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT DROP [0:0]
:DOCKER - [0:0]
:DOCKER-ISOLATION - [0:0]
-A INPUT -i lo -m comment --comment "001 accept incoming local6" -j ACCEPT
-A INPUT -p ipv6-icmp -m comment --comment "010 accept incoming ipv6-icmp" -j ACCEPT
-A INPUT -p udp -m multiport --ports 33434:33600 -m comment --comment "011 accept incoming Traceroute6" -j ACCEPT
-A INPUT -p tcp -m multiport --ports 22 -m comment --comment "012 accept incoming SSH6" -j ACCEPT
-A INPUT -m comment --comment "020 accept incoming related established rules6" -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -p tcp -m multiport --ports 80,443 -m comment --comment "040 accept incoming HTTP6" -j ACCEPT
-A INPUT -m comment --comment "998 log input6" -j LOG --log-prefix "iptables reject input: "
-A INPUT -m comment --comment "999 reject input6" -j REJECT --reject-with icmp6-port-unreachable
-A FORWARD -d fd4d:6169:6c63:6f77::/64 -i eth0 -p tcp -m multiport --ports 25,110,143,465,587,993,995 -m comment --comment "020 accept mail6" -j ACCEPT
-A FORWARD -d fc00::/7 -m comment --comment "040 docker bridged forward in6" -j ACCEPT
-A FORWARD -s fc00::/7 -m comment --comment "040 docker bridged forward out6" -j ACCEPT
-A FORWARD -s fc00::/7 -i docker0 -o eth0 -m comment --comment "040 docker0 to eth06" -j ACCEPT
-A FORWARD -p ipv6-icmp -m comment --comment "040 dokcer bridgend forward ICMP" -j ACCEPT
-A FORWARD -d fc00::/7 -i eth0 -o docker0 -m comment --comment "040 eth0 to docker06" -j ACCEPT
-A FORWARD -m comment --comment "998 log forward6" -j LOG --log-prefix "iptables reject forward: "
-A FORWARD -m comment --comment "998 reject forward6" -j REJECT --reject-with icmp6-port-unreachable
-A FORWARD -i br-938897f8afb4 ! -o br-938897f8afb4 -j ACCEPT
-A OUTPUT -o lo -m comment --comment "001 accept outgoing local6" -j ACCEPT
-A OUTPUT -p ipv6-icmp -m comment --comment "010 accept outgoing ipv6-icmp" -j ACCEPT
-A OUTPUT -p udp -m multiport --ports 33434:33600 -m comment --comment "011 accept outgoing Traceroute6" -j ACCEPT
-A OUTPUT -p tcp -m multiport --ports 22 -m comment --comment "012 accept outgoing SSH6" -j ACCEPT
-A OUTPUT -m comment --comment "020 accept outging related established rules6" -m state --state RELATED,ESTABLISHED -j ACCEPT
-A OUTPUT -d 2a04:503:0:102::39/128 -p tcp -m multiport --ports 443 -m comment --comment "021 accept HTTPS to apt.snowflakehosting.ch6" -j ACCEPT
-A OUTPUT -s fc00::/7 -m comment --comment "040 docker bridged output6" -j ACCEPT
-A OUTPUT -s fc00::/7 -o docker0 -m comment --comment "040 docker local6" -j ACCEPT
-A OUTPUT -d 2a04:503:0:105::19/128 -p udp -m multiport --ports 123 -m comment --comment "081 accept NTP to time.snowflakehosting.ch6" -j ACCEPT
-A OUTPUT -m comment --comment "998 log output6" -j LOG --log-prefix "iptables reject output: "
-A OUTPUT -m comment --comment "998 reject output6" -j REJECT --reject-with icmp6-port-unreachable
-A DOCKER -d fd4d:6169:6c63:6f77::7/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 995 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::7/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 110 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::7/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 143 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::7/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 4190 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::7/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 993 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 587 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 25 -j ACCEPT
-A DOCKER -d fd4d:6169:6c63:6f77::9/128 ! -i br-938897f8afb4 -o br-938897f8afb4 -p tcp -m tcp --dport 465 -j ACCEPT
-A DOCKER-ISOLATION -j RETURN
COMMIT
# Completed on Fri Jun  9 20:53:04 2017
```

**Comment by niklasdoerfler** [comment]
Thanks for your help! Your config looks very similar to mine, but it didn't work anyway.
But after adding some logging to the iptable rules I found out, that incoming traffic on the docker bridge (_br-dc3d3dbabd82_) was blocked. Then I added following rule to allow those incoming connections:

`-A INPUT -i br-dc3d3dbabd82 -j ACCEPT`

For the moment it seems like it's working now! :)

But in general this iptables rule should be created automatically from docker, because the bridge name changes, everytime the docker network is restarted. Is it possible to fix this?
